--- a/pyproject.toml	2026-08-07 17:12:00.000000000 -0300
+++ b/pyproject.toml	2026-08-17 14:08:24.606094280 -0300
@@ -7,7 +7,6 @@
 build-backend = "setuptools.build_meta"
 requires = [
   "setuptools>=45",
-  "packaging>=26.2",
   'typing-extensions>=4.1; python_version < "3.11"',
 ]
 
@@ -38,7 +37,6 @@
   "version",
 ]
 dependencies = [
-  "packaging>=26.2",
   'tomli>=1; python_version < "3.11"',
   'typing-extensions>=4.1; python_version < "3.11"',
 ]
