--- a/lib/replace/wscript
+++ b/lib/replace/wscript
@@ -37,7 +37,7 @@ def configure(conf):
     conf.CHECK_HEADERS('fcntl.h fnmatch.h glob.h history.h krb5.h langinfo.h')
     conf.CHECK_HEADERS('libaio.h locale.h ndir.h pwd.h')
     conf.CHECK_HEADERS('shadow.h sys/acl.h')
-    conf.CHECK_HEADERS('sys/attributes.h attr/attributes.h sys/capability.h sys/dir.h sys/epoll.h')
+    conf.CHECK_HEADERS('sys/attributes.h sys/capability.h sys/dir.h sys/epoll.h')
     conf.CHECK_HEADERS('port.h')
     conf.CHECK_HEADERS('sys/fcntl.h sys/filio.h sys/filsys.h sys/fs/s5param.h sys/fs/vx/quota.h')
     conf.CHECK_HEADERS('sys/id.h sys/ioctl.h sys/ipc.h sys/mman.h sys/mode.h sys/ndir.h sys/priv.h')
@@ -311,23 +311,23 @@ def configure(conf):
 
     xattr_headers='sys/attributes.h attr/xattr.h sys/xattr.h'
 
-    conf.CHECK_FUNCS_IN('''
+    conf.CHECK_FUNCS('''
 fgetxattr flistea flistxattr
 fremovexattr fsetxattr getxattr
 listxattr removexattr setxattr
-''', 'attr', checklibc=True, headers=xattr_headers)
+''', headers=xattr_headers)
 
     # We need to check for linux xattrs first, as we do not wish to link to -lattr
     # (the XFS compat API) on Linux systems with the native xattr API
     if not conf.CONFIG_SET('HAVE_GETXATTR'):
-        conf.CHECK_FUNCS_IN('''
+        conf.CHECK_FUNCS('''
 attr_get attr_getf attr_list attr_listf attropen attr_remove
 attr_removef attr_set attr_setf extattr_delete_fd extattr_delete_file
 extattr_get_fd extattr_get_file extattr_list_fd extattr_list_file
 extattr_set_fd extattr_set_file fgetea
 fremoveea fsetea getea listea
 removeea setea
-''', 'attr', checklibc=True, headers=xattr_headers)
+''', headers=xattr_headers)
 
     if (conf.CONFIG_SET('HAVE_ATTR_LISTF') or
         conf.CONFIG_SET('HAVE_EXTATTR_LIST_FD') or
