--- a/config/hwloc_internal.m4
+++ b/config/hwloc_internal.m4
@@ -283,36 +283,36 @@ EOF
     AC_CHECK_HEADERS([langinfo.h], [
       AC_CHECK_FUNCS([nl_langinfo])
     ])
-    hwloc_old_LIBS="$LIBS"
-    chosen_curses=""
-    for curses in ncurses curses
-    do
-      for lib in "" -ltermcap -l${curses}w -l$curses
-      do
-        AC_MSG_CHECKING(termcap support using $curses and $lib)
-        LIBS="$hwloc_old_LIBS $lib"
-        AC_LINK_IFELSE([AC_LANG_PROGRAM([[
-#include <$curses.h>
-#include <term.h>
-]], [[tparm(NULL, 0, 0, 0, 0, 0, 0, 0, 0, 0)]])], [
-          AC_MSG_RESULT(yes)
-          AC_SUBST([HWLOC_TERMCAP_LIBS], ["$LIBS"])
-          AC_DEFINE([HWLOC_HAVE_LIBTERMCAP], [1],
-                    [Define to 1 if you have a library providing the termcap interface])
-          chosen_curses=$curses
-        ], [
-          AC_MSG_RESULT(no)
-        ])
-        test "x$chosen_curses" != "x" && break
-      done
-      test "x$chosen_curses" != "x" && break
-    done
-    if test "$chosen_curses" = ncurses
-    then
-      AC_DEFINE([HWLOC_USE_NCURSES], [1], [Define to 1 if ncurses works, preferred over curses])
-    fi
-    LIBS="$hwloc_old_LIBS"
-    unset hwloc_old_LIBS
+#    hwloc_old_LIBS="$LIBS"
+#    chosen_curses=""
+#    for curses in ncurses curses
+#    do
+#      for lib in "" -ltermcap -l${curses}w -l$curses
+#      do
+#        AC_MSG_CHECKING(termcap support using $curses and $lib)
+#        LIBS="$hwloc_old_LIBS $lib"
+#        AC_LINK_IFELSE([AC_LANG_PROGRAM([[
+##include <$curses.h>
+##include <term.h>
+#]], [[tparm(NULL, 0, 0, 0, 0, 0, 0, 0, 0, 0)]])], [
+#          AC_MSG_RESULT(yes)
+#          AC_SUBST([HWLOC_TERMCAP_LIBS], ["$LIBS"])
+#          AC_DEFINE([HWLOC_HAVE_LIBTERMCAP], [1],
+#                    [Define to 1 if you have a library providing the termcap interface])
+#          chosen_curses=$curses
+#        ], [
+#          AC_MSG_RESULT(no)
+#        ])
+#        test "x$chosen_curses" != "x" && break
+#      done
+#      test "x$chosen_curses" != "x" && break
+#    done
+#    if test "$chosen_curses" = ncurses
+#    then
+#      AC_DEFINE([HWLOC_USE_NCURSES], [1], [Define to 1 if ncurses works, preferred over curses])
+#    fi
+#    LIBS="$hwloc_old_LIBS"
+#    unset hwloc_old_LIBS
 
     _HWLOC_CHECK_DIFF_U
     _HWLOC_CHECK_DIFF_W
