--- a/src/ext-fwd.c
+++ b/src/ext-fwd.c
@@ -210,7 +210,7 @@ void fwd_handle(int _rc, int _sock)
 		} else if (rc == PF_RETRY) {
 			// return;
 		} else {
-			log_err("FWD: Unhandled NAT-PMP reply.");
+			log_error("FWD: Unhandled NAT-PMP reply.");
 		}
 	}
 #endif
@@ -234,7 +234,7 @@ void fwd_handle(int _rc, int _sock)
 		} else if (rc == PF_RETRY) {
 			//return;
 		} else {
-			log_err("FWD: Unhandled UPnP reply.");
+			log_error("FWD: Unhandled UPnP reply.");
 		}
 	}
 #endif
