--- a/Modules/_ssl.c
+++ b/Modules/_ssl.c
@@ -200,7 +200,6 @@ static void _PySSLFixErrno(void) {
 #define TLS_server_method SSLv23_server_method
 #define X509_get0_notBefore X509_get_notBefore
 #define X509_get0_notAfter X509_get_notAfter
-#define ASN1_STRING_get0_data ASN1_STRING_data
 #define OpenSSL_version_num SSLeay
 #define OpenSSL_version SSLeay_version
 #define OPENSSL_VERSION SSLEAY_VERSION
@@ -860,7 +859,7 @@ _ssl_configure_hostname(PySSLSocket *sel
                 goto error;
             }
         } else {
-            if (!X509_VERIFY_PARAM_set1_ip(param, ASN1_STRING_data(ip),
+            if (!X509_VERIFY_PARAM_set1_ip(param, ASN1_STRING_get0_data(ip),
                                            ASN1_STRING_length(ip))) {
                 _setSSLError(NULL, 0, __FILE__, __LINE__);
                 goto error;
