--- a/uwsgiconfig.py
+++ b/uwsgiconfig.py
@@ -1105,10 +1105,10 @@ class uConf(object):
                 report['routing'] = True
 
 
-        if self.has_include('sys/capability.h') and uwsgi_os == 'Linux':
-            self.cflags.append("-DUWSGI_CAP")
-            self.libs.append('-lcap')
-            report['capabilities'] = True
+#        if self.has_include('sys/capability.h') and uwsgi_os == 'Linux':
+#            self.cflags.append("-DUWSGI_CAP")
+#            self.libs.append('-lcap')
+#            report['capabilities'] = True
 
         if self.has_include('uuid/uuid.h'):
             self.cflags.append("-DUWSGI_UUID")
