--- a/src/libudev.h
+++ b/src/libudev.h
@@ -11,6 +11,11 @@
 #include <fcntl.h>
 #include <unistd.h>
 
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
 struct udev;
 struct udev_device;
 struct udev_list_entry;
@@ -78,4 +83,8 @@ struct udev_device *udev_monitor_receive
     struct udev_monitor *udev_monitor);
 void udev_monitor_unref(struct udev_monitor *udev_monitor);
 
+#ifdef __cplusplus
+} /* extern "C" */
+#endif
+
 #endif
