--- a/configure
+++ b/configure
@@ -5255,8 +5255,8 @@ elif enabled mips; then
         disable mips64r6
         disable loongson2
         disable loongson3
-        disable mipsdsp
-        disable mipsdspr2
+#        disable mipsdsp
+#        disable mipsdspr2
 
         cpuflags="-march=$cpu"
 
--- a/libavcodec/mips/ac3dsp_mips.c
+++ b/libavcodec/mips/ac3dsp_mips.c
@@ -54,9 +54,11 @@
  */
 
 #include "config.h"
+#include <string.h>
 #include "libavcodec/ac3dsp.h"
 #include "libavcodec/ac3.h"
 #include "libavcodec/ac3tab.h"
+#include "libavutil/macros.h"
 #include "libavutil/mips/asmdefs.h"
 
 #if HAVE_INLINE_ASM
