From 12589d85c26258e65d817ea92d3d0b07a5feed7c Mon Sep 17 00:00:00 2001
From: Willem Toorop <willem@nlnetlabs.nl>
Date: Tue, 12 Jun 2018 17:00:45 +0200
Subject: [PATCH] Wild guess at OpenSSL without engine support

---
 src/gldns/keyraw.c | 16 +++++++++++++++-
 1 file changed, 15 insertions(+), 1 deletion(-)

diff --git a/src/gldns/keyraw.c b/src/gldns/keyraw.c
index ed8188c..7fa5bc1 100644
--- a/src/gldns/keyraw.c
+++ b/src/gldns/keyraw.c
@@ -20,8 +20,11 @@
 #include <openssl/rand.h>
 #include <openssl/err.h>
 #include <openssl/md5.h>
+#ifdef HAVE_OPENSSL_CONF_H
+# include <openssl/conf.h>
+#endif
 #ifdef HAVE_OPENSSL_ENGINE_H
-#  include <openssl/engine.h>
+# include <openssl/engine.h>
 #endif
 #ifdef HAVE_OPENSSL_BN_H
 #include <openssl/bn.h>
@@ -122,6 +125,16 @@ uint16_t gldns_calc_keytag_raw(const uint8_t* key, size_t keysize)
 #ifdef HAVE_SSL
 #ifdef USE_GOST
 /** store GOST engine reference loaded into OpenSSL library */
+#ifdef OPENSSL_NO_ENGINE
+int
+gldns_key_EVP_load_gost_id(void)
+{
+	return 0;
+}
+void gldns_key_EVP_unload_gost(void)
+{
+}
+#else
 ENGINE* gldns_gost_engine = NULL;
 
 int
@@ -181,6 +194,7 @@ void gldns_key_EVP_unload_gost(void)
                 gldns_gost_engine = NULL;
         }
 }
+#endif /* ifndef OPENSSL_NO_ENGINE */
 #endif /* USE_GOST */
 
 DSA *
-- 
2.17.1

