--- a/configure
+++ b/configure
@@ -18718,7 +18718,7 @@ else
 
     vuln=`grep "ZLIB_VERSION \"1.2.0" $ZLIB_HOME/include/zlib.h`
     if test -z "$vuln"; then
-	vuln=`grep "ZLIB_VERSION \"1.2.1" $ZLIB_HOME/include/zlib.h`
+	vuln=`grep "ZLIB_VERSION \"1.2.1[^0-9]" $ZLIB_HOME/include/zlib.h`
     fi
 
     if test -n "$vuln"; then
--- a/m4/reorganization/libs/libz.m4
+++ b/m4/reorganization/libs/libz.m4
@@ -31,7 +31,7 @@ else
 
     vuln=`grep "ZLIB_VERSION \"1.2.0" $ZLIB_HOME/include/zlib.h`
     if test -z "$vuln"; then
-	vuln=`grep "ZLIB_VERSION \"1.2.1" $ZLIB_HOME/include/zlib.h`
+	vuln=`grep "ZLIB_VERSION \"1.2.1[^0-9]" $ZLIB_HOME/include/zlib.h`
     fi
 
     if test -n "$vuln"; then
