--- a/libi2pd/Reseed.cpp
+++ b/libi2pd/Reseed.cpp
@@ -522,7 +522,11 @@ namespace data
 		boost::asio::io_service service;
 		boost::system::error_code ecode;
 
+#if BOOST_VERSION >= 104700
+		boost::asio::ssl::context ctx(boost::asio::ssl::context::sslv23);
+#else
 		boost::asio::ssl::context ctx(service, boost::asio::ssl::context::sslv23);
+#endif
 		ctx.set_verify_mode(boost::asio::ssl::context::verify_none);
 		boost::asio::ssl::stream<boost::asio::ip::tcp::socket> s(service, ctx);
 
--- a/daemon/I2PControl.cpp
+++ b/daemon/I2PControl.cpp
@@ -35,7 +35,11 @@ namespace client
 	I2PControlService::I2PControlService (const std::string& address, int port):
 		m_IsRunning (false), m_Thread (nullptr),
 		m_Acceptor (m_Service, boost::asio::ip::tcp::endpoint(boost::asio::ip::address::from_string(address), port)),
+#if BOOST_VERSION >= 104700
+		m_SSLContext (boost::asio::ssl::context::sslv23),
+#else
 		m_SSLContext (m_Service, boost::asio::ssl::context::sslv23),
+#endif
 		m_ShutdownTimer (m_Service)
 	{
 		i2p::config::GetOption("i2pcontrol.password", m_Password);
